support/nfs/cacheio.c

```diff
@@ -258,3 +260,137 @@ cache_flush(int force)
 		}
 	}
 }
+
+/*
+ * Update the expiration of an existing entry in cache so that it gets evicted.
+ * cache is either 'nfsd.fh' or 'nfsd.export'
+ */
+static void
+cache_update_entry(const char *cache, const char *message)
+{
+	char path[MAXPATHLEN];
+	int fd;
+	ssize_t mesg_len = (ssize_t)strlen(message);
+
+	snprintf(path, sizeof (path), "/proc/net/rpc/%s/channel", cache);
+	fd = open(path, O_RDWR);
+	if (fd >= 0) {
+		if (write(fd, message, mesg_len) != mesg_len) {
+			xlog_warn("Writing '%s' to '%s' failed: errno %d (%s)",
+			message, path, errno, strerror(errno));
+		}
+		close(fd);
+	}
+}
+
+/*
+ * Check if there is an entry in cache for mountpoint.
+ * cache is either 'nfsd.fh' or 'nfsd.export'
+ */
+static bool
+cache_has_entry(const char *cache, const char *mountpoint)
+{
+	char path[MAXPATHLEN];
+	bool found = false;
+	FILE *fp;
+
+	snprintf(path, sizeof (path), "/proc/net/rpc/%s/content", cache);
+	fp = fopen(path, "r");
+	if (fp != NULL) {
+		char *linebuf = NULL;
+		size_t linelen = 0;
+
+		while(getline(&linebuf, &linelen, fp) != -1) {
+			if (linebuf[0] == '#')
+				continue;
+			if (strstr(linebuf, mountpoint) != NULL) {
+				found = true;
+				break;
+			}
+		}
+		free(linebuf);
+		(void) fclose(fp);
+	}
+
+	return found;
+}
+
+#define MILLISEC        1000
+#define NANOSEC         1000000000
+#define NSEC2MSEC(n)    ((n) / (NANOSEC / MILLISEC))
+
+static uint64_t
+gethrtime(void)
+{
+        struct timespec ts;
+        (void) clock_gettime(CLOCK_MONOTONIC, &ts);
+        return ((((uint64_t)ts.tv_sec) * NANOSEC) + ts.tv_nsec);
+}
+
+static bool
+can_umount(const char *path)
+{
+	struct statfs64 stfs;
+	int error = 0;
+
+	/* With MNT_EXPIRE, umount2() will return EBUSY or EAGAIN (not busy) */
+	if (umount2(path, MNT_EXPIRE) == -1)
+		error = errno;
+
+	/* Make a call to clear our mnt_expiry_mark on the mount */
+	(void) statfs64(path, &stfs);
+
+	return (error != EBUSY);
```

Comment:
Question: doesn't the first call using the `MNT_EXPIRE` flag return `EAGAIN` if the mount is not in use, marking the mount as expired. It is then a subsequent call to `umount2()` with `MNT_EXPIRE` that will umount the mount? If that's true then for an unbusy mount we will exit the driving loop before the second `unmount2()`? I suspect the mentioned `nfsd_filecache_laundrette` thread is in play with this logic, just looking for clarification, thank you.

Reply:
I updated the comments. Basically we are leveraging the `MNT_EXPIRE` mechanism to tell us if a mount is still busy (so we can keep waiting for the file cache to drain).  We undo the expired state by making a `statfs64()` call to the mount.

Reply:
Ah, I understand now; comments cleared this up, thanks.